pkg/reconciler/namespace/namespace.go

```diff
@@ -187,111 +185,31 @@ func (c *Controller) updateScheduledCondition(ctx context.Context, ns *corev1.Na
-			klog.V(6).InfoS("assignCluster: excluding cluster with different metadata.clusterName", "ns.clusterName", ns.ClusterName, "check", allClusters[i].ClusterName)
-			continue
-		}
-		if allClusters[i].Spec.Unschedulable {
-			klog.V(2).InfoS("assignCluster: excluding unschedulable cluster", "metadata.name", allClusters[i].Name)
-			continue
-		}
-		if evictAfter := allClusters[i].Spec.EvictAfter; evictAfter != nil && evictAfter.Time.Before(time.Now()) {
-			klog.V(2).InfoS("assignCluster: excluding cluster with evictAfter value that has passed", "metadata.name", allClusters[i].Name)
-			continue
-		}
-		if !conditions.IsTrue(allClusters[i], clusterv1alpha1.ClusterReadyCondition) {
-			klog.V(2).InfoS("assignCluster: excluding not-ready cluster", "metadata.name", allClusters[i].Name)
-			continue
-		}
-
-		klog.V(2).InfoS("assignCluster: found a ready candidate", "metadata.name", allClusters[i].Name)
-		clusters = append(clusters, allClusters[i])
-	}
-
-	newClusterName := ""
-	if len(clusters) > 0 {
-		// Select a cluster at random.
-		cluster := clusters[rand.Intn(len(clusters))]
-		newClusterName = cluster.Name
-	}
-
+// ensureScheduled attempts to ensure the namespace is assigned to a viable cluster. This
+// will succeed without error if a cluster is assigned or if there are no viable clusters
+// to assign to. The condition of not being scheduled to a cluster will be reflected in
+// the namespace's status rather than by returning an error.
+func (c *Controller) ensureScheduled(ctx context.Context, ns *corev1.Namespace) error {
 	oldClusterName := ns.Labels[clusterLabel]
-	if oldClusterName != newClusterName {
-		patchType, patchBytes := ClusterLabelPatchBytes(newClusterName)
-		if _, err = c.kubeClient.Cluster(ns.ClusterName).CoreV1().Namespaces().
-			Patch(ctx, ns.Name, patchType, patchBytes, metav1.PatchOptions{}); err != nil {
-			return err
-		}
-		klog.Infof("Patched cluster assignment for namespace %q|%q: %q -> %q", ns.ClusterName, ns.Name, oldClusterName, newClusterName)
-	}
 
-	return nil
-}
-
-// isValidCluster checks whether the given cluster name exists and is valid for
-// the purposes of any namespace already scheduled to it (i.e., if it reports
-// as Ready, and any evictAfter value, if specified, has not yet passed).
-//
-// It doesn't take into account Unschedulable, and should only be used when
-// determining if a cluster that a namespace has already been assigned to
-// should keep having that namespace.
-func (c *Controller) isValidCluster(ctx context.Context, lclusterName, clusterName string) (bool, error) {
-	cluster, err := c.clusterLister.Get(clusters.ToClusterAwareKey(lclusterName, clusterName))
-	if apierrors.IsNotFound(err) {
-		klog.Infof("Cluster %q|%q does not exist", lclusterName, clusterName)
-		return false, nil
+	scheduler := namespaceScheduler{
+		getCluster:   c.clusterLister.Get,
+		listClusters: c.clusterLister.List,
 	}
+	newClusterName, err := scheduler.AssignCluster(ns)
```

Comment:
newPClusterName

Reply:
Maybe that kind of rename can wait for a separate PR?